### DISCUSSION TITLE
Support for 32-bit hardware platforms?

### DISCUSSION THREAD

**Discussion by JakubSosnovec** [eclipse-iceoryx/iceoryx#1769] [Ideas]
I understand that officially, iceoryx only supports 64-bit hardware platforms. I know about the requirement that the platform supports lock-free 64-bit atomics.

- Is there any other strong reason why 32-bit platforms cannot be supported?
- Can you expand on the lock-free 64-bit atomics requirement? Why is it needed?

In our experiments, we commented out the 64-bit atomics check (https://github.com/eclipse-iceoryx/iceoryx/blob/master/iceoryx_hoofs/source/concurrent/loffli.cpp#L32) and we then we can actually build for 32-bit platforms. Furthermore, iceoryx seems to run correctly, as far as our testing goes (which isnt too far at all).

However, we get a ton of compiler warnings. However, these are just two types:

- Wconversion: when using 64-bit values to index into pointers and arrays, etc. It seems that all integer-templated types in iceoryx_hoofs use `uint64_t` for the template parameter, for example the `iox::cxx::vector` class declaration looks as: `template <typename T, uint64_t Capacity> class vector`. Furthermore, for example even the index operator of `iox::cxx::vector` takes a `uint64_t` parameter. But, in C++, AFAIK, indices are either `std::size_t` or `std::ptrdiff_t`, which are 32-bit integers on 32-bit integers platforms. The compiler doesnt like that.
- Wcast-align: it also complains about a lof of code in `iox::cxx::variant`, somehow it cannot recognize that the `reinterpret_cast`s there are fine because of the `alignas` specifier. Not sure why it works fine on 64-bit platforms and doesnt on 32-bit platforms, actually.

Of course, we could just disable these warnings and go about our day. But I was wondering whether there are any concerns, ideas or comments about this?

For context, we are using gcc 10.2 for the compilation. And iceoryx release v2.0.2.

**Comment by elfenpiff** [comment]
@JakubSosnovec 

> Is there any other strong reason why 32-bit platforms cannot be supported?

Actually there is no other strong reason and 2 years ago we ported iceoryx to 32-bit and it was running with the atomic restriction. But we did not pursued it any further since we had no use case at that time.

> Can you expand on the lock-free 64-bit atomics requirement? Why is it needed?

We require lock-free algorithms to ensure a thread-safe access to the shared-memory constructs. We decided against mutexs since it is possible that an application dies while holding a lock and a mutex can only be unlocked by the owner. Such a case would then lead to a total deadlock of the whole system.

Usually we use 64-bit in a way that 32-bit are some kind of index and the remaining 32-bit are an aba counter. But in most cases we could reduce the index and aba size. The lock-free algorithms are then maybe less safe but for a first shot it would be fine.
Another approach would be to deactivate the enforcement of lock-free algorithms on 32-bit. When this would be fine for you we could go that route.

> Wconversion

You are right and we decided against them. The main reason is serialization and a memory layout which is platform independent. Just think of the problem one encounters when serializing a `size_t` from 32-bit to 64-bit a vice versa.
This gets even worse. `size_t` is not necessarily 64-bit in size on a 64-bit system, it is defined as a number which can index the whole memory and on a 64-bit system with 1gb of memory 32-bit is enough.
So we stick with the types with a specific size. Additionally, when certifying the code we have to follow rules which require us to only use sized types.

As solution I would introduce a little helper in iceoryx_platform which performs a conditional `static_cast` whenever the index size does not fit. Then the warning should go away.

> Wcast-align

This is weird and has to be investigated further. I think if we want to make 32-bit fly we require a CI target for this. As soon as I have time I would dig into it but only if you are not digging faster ;)

**Reply by JakubSosnovec** [reply]
@elfenpiff 
Thanks a lot for the detailed answer!

So, I am trying to think this through. If a `std::atomic<uint64_t>` is not lock-free, that means that there is an associated mutex somewhere. However, that mutex seems to not be directly in the memory of the object itself - I found this SO question explaining how it works: https://stackoverflow.com/questions/50298358/where-is-the-lock-for-a-stdatomic It also mentions that only lock-free atomics are safe to put into shared memory segments, since each process has it own hashtable with associated mutexes. So, this seems to be a very good reason to require that `std::atomic<uint64_t>` is lock-free! :)

However, in our case, we are running FreeRTOS on a 32-bit platform. So, there are no processes at all, just tasks (basically the same as threads), and no shared memory segments, just a uniform memory address space. (We simulate shared memory segments just by allocating the buffers in the heap.) Therefore, in our case it is fine to have non-lock-free atomics, since there is just a single hashtable of compiler-generated mutexes. Hopefully I am not missing anything.

Alright, this is quite intricate, but hopefully I have a good understanding now.

> As solution I would introduce a little helper in iceoryx_platform which performs a conditional static_cast whenever the index size does not fit. Then the warning should go away.

Yeah, this sounds like a good solution for us. However, you would need to put this helper function around every low-level indexing, right? So thats quite a lot of changes and a bit of boilerplate. But, if you are willing to put that into iceoryx, that would be good for us :) otherwise, we can surely just disable the conversion warnings, if we know there is nothing more worrying happening.

I will try to look into Wcast-align, too.

Thanks a lot again for your helpful answer!

**Reply by JakubSosnovec** [reply]
@elfenpiff Alright, I found the reason for the Wcast-align stuff. Basically, everything is said in this SO: https://stackoverflow.com/questions/25762173/why-does-wcast-align-not-warn-about-cast-from-char-to-int-on-x86

It doesnt generate any warnings on x86_64 or even arm64 (aarch64) because unaligned memory accesses are actually acceptable on those platforms. The compiler doesnt know that the `reinterpret_cast`s in `iox::cxx::variant` are actually casting a properly `alignas`ed memory, but it doesnt care becase even if it isnt aligned, that is fine, the CPU can deal with that. However, on the arm V7 platform, that seems to be a problem. It doesnt actually matter whether the platform is 32-bit or 64-bit, all that matters is whether the platform supports mis-aligned memory accesses or not. The compiler knows that and it generates warnings according to it.

The SO question mentions a new gcc option -Wcast-align=strict, which makes it produce alignment warnings even if the platform supports it. With -Wcast-align=strict, we get compiler warnings even when building iceoryx for x86_64.

I suspect that in your work around quality, safety etc., you will have to look into this anyway. I suspect that an AUTOSAR C++14 coding guidelines checker will report some violations about this. Actually, I think that `reinterpret_cast`s are completely forbidden there, so this will be an even bigger issue :)

I think that for us, we can just disable these warnings, now that we understand what is happening there. Of course, if we move to any more serious usage, we might have to revisit this again in the future.

Alternatively, it might be quite simple to modify the `iox::cxx::variant` implementation to avoid these problems. Perhaps the `call_at_index` methods in variant_internal.hpp could take a pointer directly to the `alignas`ed array of `byte_t` instead of a plain `byte_t*`. I might actually go and try to do that, if that sounds okey to you guys?

Thanks a lot of for the clarifications and help.

**Reply by elBoberido** [reply]
@JakubSosnovec I think we could even use `std::aligned_storage` like we do in the `cxx::optional` nowadays. You would just need to define an alias to pass it to these methods.

cc @elfenpiff @FerdinandSpitzschnueffler

**Reply by JakubSosnovec** [reply]
@elBoberido FYI, `std::aligned_storage` is depracted and the recommended way is now actually to use `alignas(T) std::byte[]` exactly the as is done in `cxx::variant`. https://stackoverflow.com/questions/71828288/why-is-stdaligned-storage-to-be-deprecated-in-c23-and-what-to-use-instead

Maybe it is just a matter of opinion on it, and it is okey to use `std::aligned_storage`. I hold no strong opinion on this.

I have been kinda playing around with defining a using typedef a then taking that as parameter in the functions. Something like:

```
template <uint64_t N, typename T, typename... Targs>
struct call_at_index
{
  using aligned_storage alignas(algorithm::max(alignof(T), alignof(Targs)...)) =
      byte_t[algorithm::max(sizeof(T), sizeof(Targs)...)];
  static void destructor(const uint64_t index, aligned_storage& ptr) noexcept
  {
    ...
```

But for now, I am still getting the warnings. My compiler seems stupid and it doesnt actually want to believe me that `ptr` is now safe to cast to `T*`. Not sure why.

**Reply by elfenpiff** [reply]
@elBoberido you mentioned that `std::aligned_storage` is deprecated with C++23 so @JakubSosnovec I would go for your approach here

> Alternatively, it might be quite simple to modify the iox::cxx::variant implementation to avoid these problems. Perhaps the call_at_index methods in variant_internal.hpp could take a pointer directly to the alignased array of byte_t instead of a plain byte_t*. I might actually go and try to do that, if that sounds okey to you guys?

So please try it :smile:

**Reply by elfenpiff** [reply]
@JakubSosnovec as it seems we have to dig a little deeper there. 
Another solution would be to implement our own C++14 aligned storage :/

**Reply by elBoberido** [reply]
@JakubSosnovec hahaha, that was exactly the reason we didn't use it in the first place and also why we implemented `uninitialized_array`.

Could you check if this code would give you the warning?

```cpp
struct alignas(T...) storage_t
{
    uint8_t data[algorithm::maxVal(sizeof(T)...)];
};

storage_t m_storage;
```

A `reinterpret_cast<T&>(m_storage)` should then work without a warning

**Reply by JakubSosnovec** [reply]
@elBoberido Yes, that works

**Reply by elfenpiff** [reply]
@FerdinandSpitzschnueffler I brought you here into the loop since the `uninitialized_array` is your child. We could try the approach that @elBoberido drafted for the uninitialized array to get rid of the cast align strict warning.

**Reply by FerdinandSpitzschnueffler** [reply]
@JakubSosnovec @elfenpiff I implemented @elBoberido's approach in https://github.com/eclipse-iceoryx/iceoryx/pull/1774. Maybe you can check if there are still cast align warnings.

**Reply by elBoberido** [reply]
@FerdinandSpitzschnueffler there are still some warnings and I proposed a solution in the PR. They are not related to the `UninitializedArray` changes but the `call_at_index` structs in the `variant` code. Nothing is broken but the compiler cannot proof if and therefore there is the warning.

cc @JakubSosnovec @elfenpiff

**Comment by JakubSosnovec** [comment]
@elBoberido @elfenpiff So, I managed to persuade the compiler to stop producing warnings with `-Wcast-align=strict` for `cxx::variant`. Here is the monstrosity that I came up with: https://github.com/eclipse-iceoryx/iceoryx/commit/9e8dc3a53e2111908a2412c7bef45275dfa5e2ab

Do you guys want it? I think it makes the code a bit more correct, but at the cost of more template magic...

**Reply by JakubSosnovec** [reply]
Btw, there is still quite a lot of warnings produced with`-Wcast-align=strict`, especially from `unitialized_array`.
